### PR TITLE
Recursively Find Validator Database File In Slashing Protection Commands

### LINKED ISSUE #8500: Make Slashing Protection Export Find validator.db Automatically Inside of Wallet Directory
# 🚀 Feature Request

### Description

Currently, performing a slasher protection export or import requires the --datadir flag which needs to be a path to a folder containing a `validator.db` file. This is confusing and leads users to sometimes pass in a wrong folder and receive an empty slashing protection history.

### Describe the solution you'd like

The import/export commands should recursively look into the folder specified for a validator.db file and check if it exists, then use it for performing their operations. This way, a user can simply pass in --datadir=/walletdir and not be concerned about getting an empty response. This will improve user experience and minimize risk of someone exporting an empty slashing protection history.

### PR DESCRIPTION
**Pull request by ahadda5**
**What type of PR is this?**

 Feature

**What does this PR do? Why is it needed?**
It ensures that the validator.db is found in the specified datadir or its subdirectories  before the slashing export commences.  Want to avoid going through the process if validator.db is not there. 

**Which issues(s) does this PR fix?**
Fixes #8500

### CODE REVIEW

**Comment by ahadda5** [comment]
I need some pointers before i proceed to make further changes ExportSlashingProtectionJSONCli /ImportSlashingProtectionJSONCli 

A util func was added.   

1. It returns if a file is found in a dir or its sub  (a bool) and the path of that file.  While we do not need the path for our issue  [8500](https://github.com/prysmaticlabs/prysm/issues/8500),  i kept it for future use?! 
2. When unit testing , the found flag (bool) works however it is worth noting that once creating tmp dir the full path cannot be pre - determined  so i commented that assert. 
3. If that is fair, i can proceed to make the change to ExportSlashingProtectionJSONCli /ImportSlashingProtectionJSONCli   ensuring that validator.db exists in the dir/sub_dir and stop with an error if that is not the case. 

`//assert.DeepEqual(t, tt.path, fpath)`
```
FAIL: TestRecursiveFileFind (0.00s)
    --- FAIL: TestRecursiveFileFind/file1 (0.00s)
        assertions.go:48: fileutil_test.go:289 Values are not equal, want: "subfolder1/subfolder11/file1", got: "/tmp/TestRecursiveFileFind269727708/001/subfolder1/subfolder11/file1", diff: modified:  = "/tmp/TestRecursiveFileFind269727708/001/subfolder1/subfolder11/file1"
    --- FAIL: TestRecursiveFileFind/file2 (0.00s)
        assertions.go:48: fileutil_test.go:289 Values are not equal, want: "subfolder2/file2", got: "/tmp/TestRecursiveFileFind269727708/001/subfolder2/file2", diff: modified:  = "/tmp/TestRecursiveFileFind269727708/001/subfolder2/file2"
    --- FAIL: TestRecursiveFileFind/file1#01 (0.00s)
        assertions.go:48: fileutil_test.go:289 Values are not equal, want: "subfolder11/file1", got: "/tmp/TestRecursiveFileFind269727708/001/subfolder1/subfolder11/file1", diff: modified:  = "/tmp/TestRecursiveFileFind269727708/001/subfolder1/subfolder11/file1"
    --- FAIL: TestRecursiveFileFind/file3 (0.00s)
        assertions.go:48: fileutil_test.go:289 Values are not equal, want: "file3", got: "/tmp/TestRecursiveFileFind269727708/001/file3", diff: modified:  = "/tmp/TestRecursiveFileFind269727708/001/file3"
FAIL

```

File: shared/fileutil/fileutil_test.go [hunk 1]
```diff
@@ -240,6 +240,57 @@ func TestDirFiles(t *testing.T) {
...
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			found, _, err := fileutil.RecursiveFileFind(tt.name, tt.root)
+			require.NoError(t, err)
+
+			assert.DeepEqual(t, tt.found, found)
+			//assert.DeepEqual(t, tt.path, fpath)
```
Comment (rauljordan) [inline shared/fileutil/fileutil_test.go:289]: Remove commented code

**Review by rauljordan — COMMENTED** [review]
Looks good so far!

**Comment by rauljordan** [comment]
@ahadda5 we should make the title share more context. Something like: "Recursively Find Validator Database File In Slashing Protection Commands". Additionally, this PR just adds a utility function but does not integrate it where it is needed. Since this PR is quite small, it is OK to do both things in this single PR. Please revise at your own convenience and then we can review again

File: shared/fileutil/fileutil_test.go (hunk 1 above)
Comment (ahadda5) [reply shared/fileutil/fileutil_test.go:289]: ok will do.

**Comment by ahadda5** [comment]
Sorry what title are you referring to?  code commentary and/or the func name ? 

> @ahadda5 we should make the title share more context. Something like: "Recursively Find Validator Database File In Slashing Protection Commands". Additionally, this PR just adds a utility function but does not integrate it where it is needed. Since this PR is quite small, it is OK to do both things in this single PR. Please revise at your own convenience and then we can review again

**Comment by ahadda5** [comment]
unit tested the slashing
>bazel test //validator/slashing-protection:cli_import_export_test.go --test_output=streamed 

However i need to test the actual export/import  , _**how can i make sure the shell script is calling my build**_ (read somewhere that it downloads the latest and greatest build). 
> prysm.sh validator slashing-protection export --datadir=/home/ahaddad/eth2validators --slashing-protection-export-dir=/home/ahaddad

File: shared/fileutil/fileutil.go [hunk 2]
```diff
@@ -131,6 +131,37 @@ func FileExists(filename string) bool {
...
+	var fpath string
+	dir = filepath.Clean(dir)
+	found = false
+	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		// checks if its a file  and has the exact name as the filename
+		// need to break the walk function by using a non-fatal error
+		if !info.IsDir() && filename == info.Name() {
```
Comment (ahadda5) [inline shared/fileutil/fileutil.go:150]: is this the best way to check if its a file?

**Comment by rauljordan** [comment]
> prysm.sh validator slashing-protection export --datadir=/home/ahaddad/eth2validators --slashing-protection-export-dir=/home/ahaddad

bazel run //validator -- slashing-protection export --datadir=/home/ahaddad/eth2validators --slashing-protection-export-dir=/home/ahaddad

File: shared/fileutil/fileutil.go (hunk 2 above)
Comment (rauljordan) [reply shared/fileutil/fileutil.go:150]: Yep this is a fine approach

**Review by rauljordan — APPROVED** [review]
Another great contribution, thanks so much @ahadda5

**Comment by ahadda5** [comment]
oops @rauljordan  i didn't want you to merge it before i finished the test. 
 
Now as expected one has to specify the full path  (./ or C:) to make sure it is not read relative otherwise it errors. So 
`bazel run //validator -- slashing-protection export --datadir=/home/.eth2validators --slashing-protection-export-dir=/home/ahaddad`
Versus 
`bazel run //validator -- slashing-protection export --datadir=./home/.eth2validators --slashing-protection-export-dir=./home/ahaddad`

with the first resulting in "error finding validator database at path"   which is ok albeit added "feature"/warning. 

Secondly and more importantly, my validator.db is a juicy testnet db of 20 MB yet exporting it does not result in the slashing.json. Hmm..   

I'm just afraid this was not tested thoroughly. 
Cheers

**Comment by rauljordan** [comment]
Sorry for merging so quickly @ahadda5, we should open these concerns in your message above as an issue that should be resolved

**Pull request merged**